{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a6cecd68-ff14-43bf-a038-6b6af8813c76",
   "metadata": {},
   "source": [
    "## Getting started with Snakemake workflows for automated bioinformatics analysis.\n",
    "\n",
    "## Author: Ali Pirani\n",
    "\n",
    "## Software we're going to use\n",
    "\n",
    "We're going to be using [conda](https://conda.io/en/latest/) package manager for installation and [snakemake](https://snakemake.readthedocs.io/en/stable/) to build the workflow.\n",
    "\n",
    "If you want to run all of this on your own computer, you'll need to follow the install instructions below.\n",
    "\n",
    "You can see the full set of installed software requirements in a conda `environment.yml` file located under binder directory. (More on binder later.)\n",
    "\n",
    "You could use this install file to run everything we're doing today on your laptop, with: \n",
    "\n",
    "```\n",
    "conda env create --file binder/environment.yml -n smake\n",
    "conda activate smake\n",
    "pip install bash_kernel\n",
    "python -m bash_kernel.install\n",
    "```\n",
    "\n",
    "We'll be implementing a short read quality check and visualization pipeline, using [fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) and [multiqc](https://multiqc.info/) to demonstrate how to get started with the Snakemake workflow. \n",
    "\n",
    "Why use a workflow management tool?\n",
    "\n",
    "- Automate your pipeline.\n",
    "- Dependency management.\n",
    "- Reentrancy - start back up where you left off\n",
    "- Reusable\n",
    "- Documentation and logging\n",
    "- Portable\n",
    "- Reproducibility.\n",
    "\n",
    "There are many Snakemake implementations on Github that you choose from for your inspiration and directory management: https://github.com/pditommaso/awesome-pipeline\n",
    "\n",
    "## Agenda: Tasks we are going to do\n",
    "\n",
    "- Get comfortable with Jupyter Notebook\n",
    "- Install Bash Kernel for Jupyter Notebook.\n",
    "- Download fastq data for QC analysis.\n",
    "- Create a Snakefile to write our first Snakemake rule\n",
    "- Explore advanced features of Snakemake\n",
    "\n",
    "## Download Bash kernel for Jupyter notebooks\n",
    "\n",
    "Run this commands in your terminal and then refresh the notebook tab. (Launch terminal from Launcher)\n",
    "\n",
    "```\n",
    "pip install bash_kernel\n",
    "python -m bash_kernel.install\n",
    "```\n",
    "\n",
    "## Download some data\n",
    "\n",
    "Execute the below cell to fetch example fastq data. This will download the example fastq data to your data directory.\n",
    "\n",
    "Note: To run bash commands, select bash kernel from the top right corner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed4c7de1-8bc1-4482-ae51-69dabc315158",
   "metadata": {},
   "outputs": [],
   "source": [
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/0Hour_001_1.fq.gz > data/0Hour_001_1.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/6Hour_001_1.fq.gz > data/6Hour_001_1.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/0Hour_001_2.fq.gz > data/0Hour_001_2.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/6Hour_001_2.fq.gz > data/6Hour_001_2.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/6Hour_002_1.fq.gz > data/6Hour_002_1.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/6Hour_002_2.fq.gz > data/6Hour_002_2.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/0Hour_002_1.fq.gz > data/0Hour_002_1.fq.gz\n",
    "curl -L https://github.com/ctb/2019-snakemake-ucdavis/raw/9db09bc0b6a3469f8a0d4996d4b2995bf36e5d27/data/0Hour_002_2.fq.gz > data/0Hour_002_2.fq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41407813-3616-4c52-a9e3-47767a9122f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls data/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca7d00ad-9a1d-427c-be3c-cdd0ce9331f6",
   "metadata": {},
   "source": [
    "Lets activate the snakemake environment using conda."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd5e31fe-9913-4415-9a9a-f3e9ea3a7418",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda activate smake"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a34e66d-3eb2-4c48-b863-e5146fab9e4b",
   "metadata": {},
   "source": [
    "## Running snakemake!\n",
    "\n",
    "### Create your first Snakefile.\n",
    "\n",
    "A Snakemake workflow is defined by specifying rules in a Snakefile. By default, the main file snakemake tries to search in your working directory is a Snakefile.\n",
    "\n",
    "Rules breaks down the workflow into small steps (for example, the application of a single tool) by specifying how to create sets of output files from sets of input files. Snakemake automatically determines the dependencies between the rules by matching file names.\n",
    "\n",
    "We will define our rules of workflow in a Snakefile. \n",
    "\n",
    "Create a new text file by selecting (`File`, `New`, `Text file`) from the the top left corner menu and copy/paste this first rule:\n",
    "\n",
    "```\n",
    "rule fastqc_a_file:\n",
    "  shell:\n",
    "    \"fastqc data/0Hour_001_1.fq.gz\"\n",
    "```\n",
    "\n",
    "Save and Rename it to \"Snakefile\". (Right click on the file from the file explorer in left panel and select Rename)\n",
    "\n",
    "* the snakemake configuration file is by default called `Snakefile`\n",
    "\n",
    "Now, run snakemake:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27a0af9-ef2c-4610-bb8a-037d00ccf327",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56759a6e-a0d4-4eea-adf0-b775a3ad3bb1",
   "metadata": {},
   "source": [
    "What was the error?\n",
    "\n",
    "Lets run it with 1 core."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cae84bc0-acde-4507-84dc-bf09128b23fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d8d90e-c1ee-4215-9663-292ebdf2f272",
   "metadata": {},
   "source": [
    "You should see:\n",
    "\n",
    "```\n",
    "\n",
    "Building DAG of jobs...\n",
    "Using shell: /bin/bash\n",
    "Provided cores: 1 (use --cores to define parallelism)\n",
    "Rules claiming more threads will be scaled down.\n",
    "Job stats:\n",
    "job              count    min threads    max threads\n",
    "-------------  -------  -------------  -------------\n",
    "fastqc_a_file        1              1              1\n",
    "total                1              1              1\n",
    "\n",
    "Select jobs to execute...\n",
    "\n",
    "[Wed Nov  3 15:49:22 2021]\n",
    "rule fastqc_a_file:\n",
    "    jobid: 0\n",
    "    resources: tmpdir=/tmp\n",
    "\n",
    "Started analysis of 0Hour_001_1.fq.gz\n",
    "Approx 5% complete for 0Hour_001_1.fq.gz\n",
    "...\n",
    "Analysis complete for 0Hour_001_1.fq.gz\n",
    "[Wed Nov  3 15:49:27 2021]\n",
    "Finished job 0.\n",
    "1 of 1 steps (100%) done\n",
    "Complete log: /home/apirani/Session8_snakemake/.snakemake/log/2021-11-03T154921.245311.snakemake.log\n",
    "\n",
    "(smake) \n",
    "```\n",
    "\n",
    "### Updating the Snakefile to track inputs and outputs \n",
    "\n",
    "At the moment this is basically just a shell script with extra syntax... then what's the point?\n",
    "\n",
    "Well, shell scripts - and this snakefile, too - will rerun the command every time you run the file, even if there's no reason to do so because the file hasn't changed.\n",
    "\n",
    "This is particularly important for large or long workflows, where you're dealing with 10s to 100s of files that may take hours to days to process! It can be hard to figure out which files to rerun, but snakemake can really help you do this!\n",
    "\n",
    "It's hard to track this kind of thing in a shell script - I usually just comment out the lines I don't want to run, or break my commands up into multiple shell scripts so they don't take so long - but with snakemake, you can annotate the rule with input and output files! \n",
    "\n",
    "Input - - - -Process 1- - - Process 2- - - Process 3- - - - Output \n",
    "\n",
    "If you annotate the rule with input/output files, Snakemake will automatically check if the output file exists and makes a decision to either rerun everything, rerun only a part of the workflow or do nothing.\n",
    "\n",
    "We will see how it does that in the following section.\n",
    "\n",
    "Change your snakefile to look like this:\n",
    "\n",
    "```\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"data/0Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/0Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc data/0Hour_001_1.fq.gz\"\n",
    "```\n",
    "\n",
    "here, we've annotated the rule with the required **input** file, as well as the expected **output** files.\n",
    "\n",
    "**Question: how do we know what the output files are?**\n",
    "\n",
    "Now run:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d74c8767-f7b7-45e2-83e9-ebc8a07d83af",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c995749-3834-4586-8e7a-e9f493065100",
   "metadata": {},
   "source": [
    "You should see something like this:\n",
    "\n",
    "```\n",
    "Building DAG of jobs...\n",
    "Nothing to be done.\n",
    "Complete log: /home/apirani/Session8_snakemake/.snakemake/log/2021-11-03T155832.508508.snakemake.log\n",
    "```\n",
    "\n",
    "**Question: What happened?**\n",
    "\n",
    "snakemake looked at the files provided with output parameters, saw that the output files already existed, and figured out that it didn't need to do anything! So cool, right? To perform this in our scripts, you would have to write few lines of code.\n",
    "\n",
    "### Forcibly re-running things\n",
    "\n",
    "You can tell snakemake to run the rule no matter what with `-f`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd3d6ce2-d880-4c49-8aa1-147b5f64b0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake -f --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17fe90d6-a0a7-4a7b-b115-a085a16dec49",
   "metadata": {},
   "source": [
    "You can also remove an output file and it will automatically re-run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "944eee8e-0b01-407b-b9ff-d710edfae087",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm data/*.html\n",
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c691a7f6-fcce-4a33-af72-1b7c81403cc4",
   "metadata": {},
   "source": [
    "Note that you don't need to remove *all* the output files to rerun a command - just remove *one* of them.\n",
    "\n",
    "You can *also* update the timestamp on an *input* file, and snakemake will figure out that the output file is older than the input file, and will rerun things."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8297b7f9-03fe-4ae1-9e4a-45faa6272f4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "touch data/*.fq.gz\n",
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d9509c2-b0dd-4fb7-950f-c671e1dc176e",
   "metadata": {},
   "source": [
    "This feature of Snakemake will become important later.\n",
    "\n",
    "### Multiple rules\n",
    "\n",
    "Let's add a rule to run fastqc on a second file:\n",
    "\n",
    "```\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"data/0Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/0Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc data/0Hour_001_1.fq.gz\"\n",
    "\n",
    "rule fastqc_a_file2:\n",
    "  input:\n",
    "    \"data/6Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/6Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc data/6Hour_001_1.fq.gz\"\n",
    "```\n",
    "\n",
    "Now run snakemake again:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b532e550-d915-4dc3-874d-33aaf31fef16",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56aafd70-2d7f-45a8-86aa-b10cfc4760e3",
   "metadata": {},
   "source": [
    "**Question: What did snakemake do?**\n",
    "\n",
    "- It will do nothing. Because By default, snakemake only runs the *first* rule in a Snakefile.\n",
    "\n",
    "**Question:How can you bypass this?**\n",
    "\n",
    "- You can give a rule name on the command line, if you like, **or** you can tell snakemake what output file(s) you want.\n",
    "\n",
    "Lets do the latter and run the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a3fc155-0f8d-4f01-9cb9-2eef143ba998",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1 data/0Hour_001_1_fastqc.html data/6Hour_001_1_fastqc.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97252213-b66e-4495-83c2-615909e6190c",
   "metadata": {},
   "source": [
    "Now you should see the second fastqc command run, with the appropriate output files!\n",
    "\n",
    "Note that snakemake only runs the second rule, because it looks at the output files and sees that the first file you wanted, `0Hour_001_1_fastqc.html` already exists!\n",
    "\n",
    "Its a good feature when you are dealing with only few rules or while debugging your rules. The downside here is:\n",
    "\n",
    "* this is pretty long compared to the same shell script...\n",
    "* specifying which file or rule you want is kind of annoying...\n",
    "\n",
    "### A first refactoring: adding a better default rule\n",
    "\n",
    "Let's start refactoring (cleaning up) this Snakefile.\n",
    "\n",
    "First, let's add a rule at the top:\n",
    "\n",
    "```\n",
    "rule all:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"data/0Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/0Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc data/0Hour_001_1.fq.gz\"\n",
    "\n",
    "rule fastqc_a_file2:\n",
    "  input:\n",
    "    \"data/6Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/6Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc data/6Hour_001_1.fq.gz\"\n",
    "```\n",
    "\n",
    "This rule, by convention called `all`, is a default rule that produces all the final output files. \n",
    "\n",
    "But it's a bit weird! It's all input, and no output!\n",
    "\n",
    "Its a blank rule that gathers together all of the various output files you want the workflow to produce, and says \"hey, snakemake, I depend on all of these files for my input - make them for me!\" And then, once those files are all there, it ...does nothing.\n",
    "\n",
    "Yep, this is perfectly legal in snakemake, and it's one way to make your life easier.\n",
    "\n",
    "\n",
    "### A second refactoring: doing a bit of templating\n",
    "\n",
    "There's a lot of repetition in each of these rules. Let's collapse it down a little bit by replacing the filename in the fastqc command with a magic variable, `{input}`. The magic variable {input} means take files mentioned in the input section of the rule.\n",
    "\n",
    "```\n",
    "rule all:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"data/0Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/0Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule fastqc_a_file2:\n",
    "  input:\n",
    "    \"data/6Hour_001_1.fq.gz\"\n",
    "  output:\n",
    "    \"data/6Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "```\n",
    "\n",
    "This all works as before, but now the rule is a bit more generic and will work with any input file. Sort of. why?\n",
    "\n",
    "The input files data/0Hour_001_1.fq.gz and data/6Hour_001_1.fq.gz are still hardcoded! We dont want to hardcode anything! \n",
    "\n",
    "### Refactoring 3: templating output files, too\n",
    "\n",
    "What do I mean, sort of?\n",
    "\n",
    "Well, the output filenames ALSO depend on the input file names in some way - specifically, fastqc replace part of the filename with `_fastqc.html` and `_fastqc.zip` to make its two output files.\n",
    "\n",
    "Let's rewrite the rule using some snakemake pattern matching:\n",
    "\n",
    "```\n",
    "rule all:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule fastqc_a_file2:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "```\n",
    "\n",
    "What we've done here is tell snakemake that anytime we say we *want* a file that ends with `_fastqc.html`, it should look for a file that ends in `.fq.gz` and then run `fastqc` on it.\n",
    "\n",
    "Try running this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c7eabea-f756-4302-841c-32e79c96376f",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86814fd8-89d5-41cc-b249-d1a2dfe711da",
   "metadata": {},
   "source": [
    "We get a `AmbiguousRuleException:`! What's going on?\n",
    "\n",
    "Well, if you look at the rule above, we've given snakemake two different rules to produce the same file(s)! `fastqc_a_file` and `fastqc_a_file2` are now identical rules! snakemake doesn't like that.\n",
    "\n",
    "Let's remove one, to get a trimmer, leaner, and above all *functional* snakefile:\n",
    "\n",
    "```\n",
    "rule all:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "```\n",
    "\n",
    "and THAT should now work just fine!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31608a75-5ce2-42cf-9fa0-8327124e4814",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "753f3742-49d4-4af3-82fd-246872790b3d",
   "metadata": {},
   "source": [
    "### Adding some more files\n",
    "\n",
    "Now here's the fun bit -- if you look in the data directory, you'll see that there are actually 8 files in there. Let's modify the snakefile to run fastqc on all of them!\n",
    "\n",
    "**Question: How should we do that? (Give it a try!)**\n",
    "\n",
    "Note that the variable name in input and output does not have to be \"filename\", it can be anything as long as you're consistant.\n",
    "\n",
    "### Building out the workflow\n",
    "\n",
    "So, we've gotten fastqc sorted out. What's next?\n",
    "\n",
    "Let's add in a new rule - multiqc, to summarize our fastqc results.\n",
    "\n",
    "multiqc takes a directory name under which there are one or more fastqc reports, and then summarizes them.\n",
    "\n",
    "Run multiqc on the command line to determine what the output files look like,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f236af3e-b2e6-42e2-a7c1-3bd1ee1296d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "multiqc data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ca2d57f-4bf6-4c2c-90a2-d558e8829147",
   "metadata": {},
   "source": [
    "You can see that it creates two outputs, *multiqc_report.html* and the directory *multiqc_data/* which contains a bunch of files. \n",
    "\n",
    "Let's create a snakemake rule for this; add:\n",
    "\n",
    "```\n",
    "rule all:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule run_multiqc:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\",\n",
    "  output:\n",
    "    \"multiqc_report.html\",\n",
    "    directory(\"multiqc_data\")\n",
    "  shell:\n",
    "    \"multiqc data/\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c7515de-78a3-46a9-876f-9275a975899f",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed1748da-24bd-4aca-a691-36401bb9b0b5",
   "metadata": {},
   "source": [
    "This ...doesn't really do what we want, for a few reasons.\n",
    "\n",
    "First of all, the output of run_multiqc is not specified in the all rule so snakemake doesn't look for a rule to create this.\n",
    "\n",
    "Second of all, multiqc_report.html already exists, so snakemake won't run the rule.\n",
    "\n",
    "Let's fix the first two things first:\n",
    "\n",
    "    add multiqc_report.html to the inputs for the first rule - all.\n",
    "    then remove multiqc_report.html and re-run snakemake.\n",
    "\n",
    "Your snakefile should look like:\n",
    "\n",
    "```\n",
    "rule all:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\",\n",
    "    \"multiqc_report.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule run_multiqc:\n",
    "  input:\n",
    "    \"data/0Hour_001_1_fastqc.html\",\n",
    "    \"data/6Hour_001_1_fastqc.html\",\n",
    "  output:\n",
    "    \"multiqc_report.html\",\n",
    "    directory(\"multiqc_data\")\n",
    "  shell:\n",
    "    \"multiqc data/\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bf89e56-9970-4d8c-88da-2e6a81870635",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm multiqc_report.html\n",
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95251196-6ff6-421c-afc9-332779906029",
   "metadata": {},
   "source": [
    "There is a lot of redundancy in the code. Providing input files explicitly to the multiqc rule is not great because those are the same files we have in the all rule and to add new files we'd now have to add them to two places.\n",
    "\n",
    "We can fix the first issue by using variables.\n",
    "\n",
    "To use variables, let's make a Python list at the very top, containing all of our expected output files from fastqc:\n",
    "\n",
    "```\n",
    "fastqc_output = [\"data/0Hour_001_1_fastqc.html\", \"data/6Hour_001_1_fastqc.html\", \"data/0Hour_001_2_fastqc.html\", \"data/6Hour_001_2_fastqc.html\"]\n",
    "```\n",
    "\n",
    "and modify the all and multiqc rules to contain this list. \n",
    "\n",
    "The final snakefile looks like this:\n",
    "\n",
    "```\n",
    "fastqc_output = [\"data/0Hour_001_1_fastqc.html\", \"data/6Hour_001_1_fastqc.html\",\n",
    "  \"data/0Hour_001_2_fastqc.html\", \"data/6Hour_001_2_fastqc.html\"]\n",
    "\n",
    "rule all:\n",
    "  input:\n",
    "    fastqc_output,\n",
    "    \"multiqc_report.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule run_multiqc:\n",
    "  input:\n",
    "    fastqc_output\n",
    "  output:\n",
    "    \"multiqc_report.html\",\n",
    "    directory(\"multiqc_data\")\n",
    "  shell:\n",
    "    \"multiqc data/\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad161b97-55fd-4e07-b3c5-2b885fe88152",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af4caffd-f0bf-4fb1-b829-18b6df5236d2",
   "metadata": {},
   "source": [
    "## Refactoring this to make it slightly more concise \n",
    "\n",
    "We've got one more redundancy in this file - the fastqc_output is listed in the all rule, but you don't need it there! Why?\n",
    "\n",
    "Well, multiqc_report.html is already in the all rule, and the multiqc rule depends on fastqc_output, so fastqc_output already needs to be created to satisfy the all rule, so... specifying it in the all rule is redundant! And you can remove it!\n",
    "\n",
    "(It's not a big deal and I usually leave it in. But I wanted to talk about dependencies!)\n",
    "\n",
    "The Snakefile now looks like this:\n",
    "\n",
    "```\n",
    "fastqc_output = [\"data/0Hour_001_1_fastqc.html\", \"data/6Hour_001_1_fastqc.html\",\n",
    "  \"data/0Hour_001_2_fastqc.html\", \"data/6Hour_001_2_fastqc.html\"]\n",
    "\n",
    "rule all:\n",
    "  input:\n",
    "    \"multiqc_report.html\"\n",
    "\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule run_multiqc:\n",
    "  input:\n",
    "    fastqc_output\n",
    "  output:\n",
    "    \"multiqc_report.html\",\n",
    "    directory(\"multiqc_data\")\n",
    "  shell:\n",
    "    \"multiqc data/\"\n",
    "```\n",
    "\n",
    "and we can rerun it from scratch by doing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18cf1bc9-66c9-459c-a1a6-95c6827be048",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm data/*.html multiqc_report.html\n",
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f526f094-aa51-4bc1-8580-6c7021de61c3",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Question: what files does snakemake check in order to decide about rerunning?\n",
    "\n",
    "snakemake will compare only the dates on the very initial input files,\n",
    "and the specific output file(s) you are requesting, to decide if it\n",
    "needs to rerun the workflow.\n",
    "\n",
    "In practical terms, this means that if you just delete the\n",
    "`data/*.html` files above but leave `multiqc_report.html` around,\n",
    "snakemake won't rerun anything. You have to delete both the\n",
    "intermediaries _and_ the end output files (as we do in the previous\n",
    "section), _or_ update the raw input files, in order to force\n",
    "rerunning.\n",
    "\n",
    "([This is a feature, not a bug](https://bitbucket.org/snakemake/snakemake/issues/885/snakemake-sometimes-doesnt-run-rule-when) - it helps deal with data-intensive pipelines where the intermediate files are really big.)\n",
    "\n",
    "### Making a `clean` rule\n",
    "\n",
    "It's kind of annoying to have to delete things explicitly. Snakemake should take care of that for us. \n",
    "\n",
    "Let's add a new rule, `clean` at the top, that forces rerunning of things --\n",
    "\n",
    "``` \n",
    "fastqc_output = [\"data/0Hour_001_1_fastqc.html\", \"data/6Hour_001_1_fastqc.html\",\n",
    "  \"data/0Hour_001_2_fastqc.html\", \"data/6Hour_001_2_fastqc.html\"]\n",
    "\n",
    "rule all:\n",
    "  input:\n",
    "    fastqc_output,\n",
    "    \"multiqc_report.html\"\n",
    "\n",
    "rule clean:\n",
    "  shell:\n",
    "    \"rm -f {fastqc_output} multiqc_report.html\"\n",
    "    \n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "\n",
    "rule run_multiqc:\n",
    "  input:\n",
    "    fastqc_output\n",
    "  output:\n",
    "    \"multiqc_report.html\",\n",
    "    directory(\"multiqc_data\")\n",
    "  shell:\n",
    "    \"multiqc data/\"\n",
    "```\n",
    "\n",
    "and now try rerunning things:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ae5894e-4617-48b1-951d-dae72665c8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake -p clean --cores 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76ca91e3-93e7-4520-9294-5afc258443a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "080f1a99-113d-46cd-acb0-80b4cdc2e4fa",
   "metadata": {},
   "source": [
    "A few things to point out --\n",
    "\n",
    "* Here we see the use of variables inside a shell command, again - `{fastqc_output}` means \"replace the thing in the curly quotes with the Python values in `fastqc_output`.\n",
    "* We're using `snakemake -p` to get a printout of the commands that are run.\n",
    "\n",
    "What's particularly nice about the `clean` rule (the name is a convention, not a requirement) is that you only need to keep track of the expected output files in one or two places - the all rule, and the clean rule.\n",
    "\n",
    "## Recap\n",
    "\n",
    "So, we've put all this work into making this snakefile with its input rules and its output rules... and there are a lot of advantages to our current approach already! \n",
    "\n",
    "Let's list a few of them --\n",
    "\n",
    "- we've completely automated our analysis!\n",
    "- we can easily add new data files into fastqc and multiqc!\n",
    "- we can rerun things easily, and (even better) by default only things that need to be run will be run.\n",
    "- the snakefile is actually pretty reusable - we could drop this into a new project, and, with little effort, run all of these things on new data!\n",
    "\n",
    "# More advanced snakemake\n",
    "\n",
    "## Running things in parallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd8d7aca-94d5-43f2-84cc-21accae2719a",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake -p clean --cores 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3acdb93c-20dc-4bec-9b9e-520fdffe114f",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake -j 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc4120d5-be13-4b3a-882e-432631e9809a",
   "metadata": {},
   "source": [
    "This will run up to four things in parallel!\n",
    "\n",
    "Points to remember:\n",
    "\n",
    "* all the output is sort of smushed together... if a rule fails, it may be hard to figure out what happened. You can always just rerun snakemake.\n",
    "* still need to be careful about how much memory and processor you're using!\n",
    "* can be used on a cluster, to distribute jobs across multiple compute nodes. (This requires more work;)\n",
    "\n",
    "### Specifying software required for a rule\n",
    "\n",
    "You can specify software on a per-rule basis! This is really helpful when you have incompatible software requirements for different rules, or want to run on a cluster, or just want to pin your snakemake workflow to a specific version.\n",
    "\n",
    "For example, if you create a file `env_fastqc.yml` with the following content,\n",
    "\n",
    "```\n",
    "channels:\n",
    "  - bioconda\n",
    "  - defaults\n",
    "  - conda-forge\n",
    "dependencies:\n",
    "  - fastqc==0.11.8\n",
    "```\n",
    "\n",
    "and then change the fastqc rule to look like this:\n",
    "\n",
    "```\n",
    "rule fastqc_a_file:\n",
    "  input:\n",
    "    \"{filename}.fq.gz\"\n",
    "  output:\n",
    "    \"{filename}_fastqc.html\",\n",
    "    \"{filename}_fastqc.zip\"\n",
    "  conda:\n",
    "    \"env_fastqc.yml\"\n",
    "  shell:\n",
    "    \"fastqc {input}\"\n",
    "```\n",
    "\n",
    "you can now run snakemake like so,\n",
    "\n",
    "```\n",
    "snakemake --use-conda -j 4\n",
    "```\n",
    "and for that rule, snakemake will install just that software, with the specified version.\n",
    "\n",
    "This aids in reproducibility, in addition to the practical advantages of isolating software installs from each other.\n",
    "\n",
    "(You can also do this with docker and singularity containers, too!)\n",
    "\n",
    "### Outputting the entire workflow diagram\n",
    "\n",
    "You can visualize your workflow like so!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14f41b07-5c4f-44ec-91b5-e1d927e0100f",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --dag -j 4  | dot -Tpng > dag.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e1c67c-df4a-4b81-97c3-cf5e8d4b60a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake -p clean --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ea5c3e-beaf-49a1-b9ee-39ba6057ecfc",
   "metadata": {},
   "source": [
    "Running on a cluster:\n",
    "\n",
    "You can specify a cluster submit command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cadca3e-7521-417e-8745-9f24af73fc43",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --cluster sbatch -j 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "defb95d5-3e6d-4d7c-8637-57e80dfcd00c",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake -p clean --cores 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9c8a6b6-5af2-43aa-821e-e4883aac4e76",
   "metadata": {},
   "source": [
    "### Final thoughts - writing your own snakefile\n",
    "\n",
    "Just like scripting, or writing an R script, writing a snakefile is a kind of programming. So you'll have to do a lot of debugging.\n",
    "\n",
    "Workflows can get really complicated; for example [Samosa](https://github.com/alipirani88/samosa/blob/main/dag.svg)\n",
    "\n",
    "If you want to see some good examples of how to build nice, clean, simple looking workflows, check out [this RNAseq example](https://github.com/snakemake-workflows/rna-seq-star-deseq2).\n",
    "\n",
    "* start small, grow your snakefile!\n",
    "* DO copy and paste from this and others you find online!\n",
    "* it rarely hurts to just re-run snakemake!\n",
    "\n",
    "## Other topics to explore for code club: wildcards, adding python code, configuring cluster config and many more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79711f2c-2c18-4da5-ab7e-a64b00cc0253",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
